site-src/geps/gep-1364.md

```diff
@@ -0,0 +1,328 @@
+* Specify changes to conformance required for Condition updates
+
+## Non-Goals
+
+
+
+## Introduction
+
+Gateway API currently has a lot of issues related to status, especially that
+status is inconsistent ([#1111][1111]), that names are hard to understand ([#1110][1110]),
+Reasons aren't explained properly ([#1362][1362]), and so on.
+
+As the API has grown, the way we talk about resources has changed a lot, and some of the
+status design hasn't been updated since resources were created.
+
+So, for example,  we have GatewayClass with `Accepted`, Gateway with `Scheduled`,
+the Gateway Listeners with `Detached` (which you want to be `false`, unlike the previous
+two), and then Gateways and Gateway Listeners have `Ready`, but Route doesn't (and which
+also you want to be `true`).
+
+This document lays out large-scale changes to the way that we talk about resources,
+and the Conditions to match them. This means that there will be an unavoidable break
+in what constitutes a healthy or unhealthy resource, and there will be changes
+required for all conformant implementations to stay conformant.
+
+This level of change is not optimal, and the intent is to make this a one-off change
+that can be built upon for future resources - since there are definitely more resources
+on the way.
+
+## Background: Kubernetes API conventions and prior art on Conditions
+
+Because this GEP is mainly concerned with updating the Conditions we are setting in
+Gateway API resources' `status`, it's worth reviewing some important points about
+Conditions. (Thi information is mainly taken from the [Typical status properties][typstatus]
+section of the API conventions document.)
+
+* Conditions are a standard type used to represent arbitrary higher-level status from
+a controller.
+* They are a listMapType, a list that is enforced by the apiserver to have only
+one entry of each item, using the `type` field as a key. (So, this is effectively
+a map that looks like a list in YAML form).
+* Each has a number of fields, the most important of which for this discussion
+are `type`, `status`, `season`, and `observedGeneration`.
+  * `type` is a string value indicating the Condition type. `Accepted`, `Scheduled`,
+  and `Ready` are current examples.
+  * `status` indicates the state of the condition, and can be one of three values,
+  `true`, `false`, or `unknown`. Unkown in particular is important, because it
+  means that the controller is unable to determine the status for some reason.
+  * `reason` is a CamelCase string that is a brief description of the reason why
+  the `status` is set the way it is.
+  * `observedGeneration` is an optional field that sets what the `metadata.generation`
+  field was when the controller last saw a resource.
+* Conditions shoud describe the _current state_ of the resource at observation
+time, which means that they should be an adjective (like `Ready`), or a past-tense
+verb (like `Attached`).
```

Comment:
Would `Attaching` ever be valid as a condition? Maybe just verb or adjective is sufficient here?

Reply:
That sort of usage is specifically called out as not desirable in the API conventions, and I agree. Conditions are supposed to be level-triggered, not edge-triggered, so having a condition that says that something is happening is a bit weird. (Should `Attaching` transition to `false` once the attaching process is complete?)

Reply:
For posterity, can we cite some sources for that assertion? It would probably actually be good as a part of this status effort to document somewhere _in the code_ that we're avoiding this due to conventions, and then point to those conventions, as this will help future contributors to more quickly avoid pursuing such things (or give them a path to who they need to convince in the greater kubernetes community).

Reply:
This is almost a direct quote from the API conventions as it is, should I put a link in here to the exact line?

Reply:
The exact quote is as follows, I can just inline if it's better, but I was trying to avoid just pasting in the entire section:

---

Condition type names should describe the current observed state of the resource, rather than describing the current state transitions. This typically means that the name should be an adjective ("Ready", "OutOfDisk") or a past-tense verb ("Succeeded", "Failed") rather than a present-tense verb ("Deploying"). Intermediate states may be indicated by setting the status of the condition to Unknown.

For state transitions which take a long period of time (e.g. more than 1 minute), it is reasonable to treat the transition itself as an observed state. In these cases, the Condition (such as "Resizing") itself should not be transient, and should instead be signalled using the True/False/Unknown pattern. This allows other observers to determine the last update from the controller, whether successful or failed. In cases where the state transition is unable to complete and continued reconciliation is not feasible, the Reason and Message should be used to indicate that the transition failed.

---

I don't believe that we generally have any process that we want to handle in the way called out in the exception bullet point. Happy to be guided here on what you'd like to see.